src/coreComponents/fileIO/vtk/VTKPolyDataWriterInterface.cpp

```diff
@@ -91,6 +91,48 @@ gatherNbElementsInRegion( ElementRegionBase const & er,
   return nbElemsInRegion;
 }
 
+/*!
+ * @brief Insert the value at (i,j,k,l) inside the VTKGEOSXData object
+ * @param[in] data the pointer to the VTKGEOSXData object
+ * @param[in] x the array that contains the field
+ * @param[inout] count a counter to increment every time we insert a value
+ * @param[in] i the first index
+ * @param[in] j the second index
+ * @param[in] k the third index
+ * @param[in] l the fourth index
+ */
+template< typename T >
+typename std::enable_if_t< traits::is_array_type< T > && T::NDIM == 4 >
+insertValueFrom4DField( vtkSmartPointer< VTKGEOSXData > const data,
+                        T const & x,
+                        localIndex & count,
+                        localIndex i, localIndex j, localIndex k, localIndex l )
+{
+  data->customInsertValue( count++, x( i, j, k, l ) );
+}
+
+/*!
+ * @brief Insert the value at (i,j,k,l) inside the VTKGEOSXData object
+ * @param[in] data the pointer to the VTKGEOSXData object
+ * @param[in] x the array that contains the field
+ * @param[inout] count a counter to increment every time we insert a value
+ * @param[in] i the first index
+ * @param[in] j the second index
+ * @param[in] k the third index
+ * @param[in] l the fourth index
+ */
+template< typename T >
+typename std::enable_if_t< !traits::is_array_type< T > || T::NDIM != 4 >
+insertValueFrom4DField( vtkSmartPointer< VTKGEOSXData > const data,
+                        T const & x,
+                        localIndex & count,
+                        localIndex i, localIndex j, localIndex k, localIndex l )
+{
+  GEOSX_UNUSED_VAR( data, x, count, i, j, k, l );
+  GEOSX_ERROR( "This function should not be called" );
+}
```

Comment:
I don't think you need to handle non-array types here (and they wouldn't compile anyway, because `T::NDIM` wouldn't be a valid expression for anything other than an `ArrayView` or `ArraySlice`). So you could "simplify" this to:
```c++
template< typename T, int NDIM, int USD >
void insertValueFrom4DField( vtkSmartPointer< VTKGEOSXData > const data,
                             ArrayView< T, NDIM, USD > const & field,
                             ... )
{
  GEOSX_UNUSED_VAR( data, field, count, i, j, k, l );
  GEOSX_ERROR( "This function should not be called" );
}

template< typename T, int USD >
void insertValueFrom4DField( vtkSmartPointer< VTKGEOSXData > const data,
                             ArrayView< T, 4, USD > const & field,
                             ... )
{
  data->customInsertValue( count++, field( i, j, k, l ) );
}
```

This isn't any shorter than what you have, but the lack of `std::enable_if` is a small win for readability in my books :)
(On a side note, how I wish we could use C++17's `if constexpr`...)

Reply:
On a second look, please also remove the passing of `VTKGEOSXData` via smart pointer. This involves a reference count increment/decrement, which, depending on the implementation, might be somewhat expensive (e.g. acquiring a mutex lock). Best to just pass by reference (`VTKGEOSXData & data`).

Reply:
Done, I modified the function as suggested in your two comments above.